sdk/dotnet/Pulumi/Provider/Provider.cs

```diff
@@ -0,0 +1,1335 @@
+        }
+    }
+
+    public sealed class CheckRequest
+    {
+        public readonly string Urn;
+        public readonly ImmutableDictionary<string, PropertyValue> Olds;
+        public readonly ImmutableDictionary<string, PropertyValue> News;
+        public readonly ImmutableArray<byte> RandomSeed;
+
+        public CheckRequest(string urn, ImmutableDictionary<string, PropertyValue> olds, ImmutableDictionary<string, PropertyValue> news, ImmutableArray<byte> randomSeed)
+        {
+            Urn = urn;
+            Olds = olds;
+            News = news;
+            RandomSeed = randomSeed;
+        }
+    }
+
+    public sealed class CheckFailure
+    {
+        public string Property { get; set; }
+        public string Reason { get; set; }
+
+        public CheckFailure(string property, string reason)
+        {
+            Property = property;
+            Reason = reason;
+        }
+    }
+
+    public sealed class CheckResponse
+    {
+        public IDictionary<string, PropertyValue>? Inputs { get; set; }
+        public IList<CheckFailure>? Failures { get; set; }
+    }
+
+
+    public sealed class DiffRequest
+    {
+        public readonly string Urn;
+        public readonly string ID;
+        public readonly ImmutableDictionary<string, PropertyValue> Olds;
+        public readonly ImmutableDictionary<string, PropertyValue> News;
+        public readonly ImmutableArray<string> IgnoreChanges;
+
+        public DiffRequest(string urn, string id, ImmutableDictionary<string, PropertyValue> olds, ImmutableDictionary<string, PropertyValue> news, ImmutableArray<string> ignoreChanges)
+        {
+            Urn = urn;
+            ID = id;
+            Olds = olds;
+            News = news;
+            IgnoreChanges = ignoreChanges;
+        }
+    }
+
+    public enum PropertyDiffKind
+    {
+        Add = 0,
+        AddReplace = 1,
+        Delete = 2,
+        DeleteReplace = 3,
+        Update = 4,
+        UpdateReplace = 5,
+    }
+
+    public sealed class PropertyDiff
+    {
+        public PropertyDiffKind Kind { get; set; }
+        public bool InputDiff { get; set; }
+    }
+
+    public sealed class DiffResponse
+    {
+        public bool? Changes { get; set; }
+
+        public IList<string>? Replaces { get; set; }
+
+        public IList<string>? Stables { get; set; }
+
+        public bool DeleteBeforeReplace { get; set; }
+        public IList<string>? Diffs { get; set; }
+
+        public IDictionary<string, PropertyDiff>? DetailedDiff { get; set; }
+    }
+
+    public sealed class InvokeRequest
+    {
+        public readonly string Tok;
+        public readonly ImmutableDictionary<string, PropertyValue> Args;
+
+        public InvokeRequest(string tok, ImmutableDictionary<string, PropertyValue> args)
+        {
+            Tok = tok;
+            Args = args;
+        }
+    }
+
+    public sealed class InvokeResponse
+    {
+
+        public IDictionary<string, PropertyValue>? Return { get; set; }
```

Comment:
With the work being done in #11418, it should be possible to define an invoke response that returns a single `PropertyValue` rather than a bag of these

Reply:
Yeh I'll probably rejig this, few other things I've copied from the go SDK which are more legacy than really needed.